collection/collection_map.go

```diff
@@ -84,114 +128,61 @@ func (c *Map) FindAll() []types.MatchData {
-// AddUnique will add a value to a key if it is not already there
-func (c *Map) AddUnique(key string, value string) {
-	c.AddUniqueCS(key, key, value)
-}
-
-// SetCS will replace the key's value with this slice
-// internally converts [] string to []types.AnchoredVar
-// with case sensitive vKey
-func (c *Map) SetCS(key string, vKey string, values []string) {
-	c.data[key] = make([]types.AnchoredVar, 0, len(values))
-	for _, v := range values {
-		c.data[key] = append(c.data[key], types.AnchoredVar{Name: vKey, Value: v})
-	}
+func (c *defaultMap) Add(key string, value string) {
+	keyL := strings.ToLower(key)
+	aVal := types.AnchoredVar{Name: key, Value: value}
+	c.data[keyL] = append(c.data[keyL], aVal)
 }
 
 // Set will replace the key's value with this slice
 // internally converts [] string to []types.AnchoredVar
-func (c *Map) Set(key string, values []string) {
-	c.SetCS(key, key, values)
-}
-
-// SetIndexCS will place the value under the index
-// If the index is higher than the current size of the CollectionMap
-// it will be appended
-// with case sensitive vKey
-func (c *Map) SetIndexCS(key string, index int, vKey string, value string) {
-	if c.data[key] == nil {
-		c.data[key] = []types.AnchoredVar{{Name: vKey, Value: value}}
-	}
-	vVal := types.AnchoredVar{Name: vKey, Value: value}
-	if len(c.data[key]) <= index {
-		c.data[key] = append(c.data[key], vVal)
-		return
+func (c *defaultMap) Set(key string, values []string) {
+	keyL := strings.ToLower(key)
+	c.data[keyL] = make([]types.AnchoredVar, 0, len(values))
+	for _, v := range values {
+		c.data[keyL] = append(c.data[keyL], types.AnchoredVar{Name: key, Value: v})
 	}
-	c.data[key][index] = vVal
 }
 
 // SetIndex will place the value under the index
 // If the index is higher than the current size of the CollectionMap
 // it will be appended
-func (c *Map) SetIndex(key string, index int, value string) {
-	c.SetIndexCS(key, index, key, value)
+func (c *defaultMap) SetIndex(key string, index int, value string) {
+	keyL := strings.ToLower(key)
+	values := c.data[keyL]
+	av := types.AnchoredVar{Name: key, Value: value}
+
+	switch {
+	case len(values) == 0:
+		c.data[keyL] = []types.AnchoredVar{av}
+	case len(values) <= index:
+		c.data[keyL] = append(c.data[keyL], av)
+	default:
+		c.data[keyL][index] = av
+	}
 }
 
 // Remove deletes the key from the CollectionMap
-func (c *Map) Remove(key string) {
-	delete(c.data, key)
+func (c *defaultMap) Remove(key string) {
+	keyL := strings.ToLower(key)
+	delete(c.data, keyL)
 }
 
 // Name returns the name for the current CollectionMap
-func (c *Map) Name() string {
+func (c *defaultMap) Name() string {
 	return c.name
 }
 
 // Reset the current CollectionMap
-func (c *Map) Reset() {
+func (c *defaultMap) Reset() {
 	for k := range c.data {
 		delete(c.data, k)
 	}
 }
 
 // Data returns all the data in the CollectionMap
-func (c *Map) Data() map[string][]string {
+func (c *defaultMap) Data() map[string][]string {
 	result := map[string][]string{}
```

Comment:
if `c.data` is empty, shall we return `nil`?

Reply:
`nil` map isn't usable like a slice so think empty map is better